File: src/upros_tutorial/src/ros_subscriber.cpp
#include <ros/ros.h>
#include "upros_message/MyMessage.h"

void callback(const upros_message::MyMessage::ConstPtr &msg) {
    std::string str = "callback   key : " + std::to_string(msg->key) + " , value : " + msg->value;
    std::cout << str << std::endl;
}

int main(int argc, char **argv) {
    // 初始化ROS节点
    ros::init(argc, argv, "my_subscriber");

    ROS_INFO("start a subscriber node");

    // 创建节点句柄
    ros::NodeHandle nh;

    // 定义一个订阅者对象
    ros::Subscriber sub = nh.subscribe("my_topic", 10, callback);

    // 运行节点
    ros::spin();

    return 0;
}

File: src/upros_tutorial/scripts/ros_subscriber.py
#!/usr/bin/env python3

import rospy
from upros_message.msg import MyMessage


#定义一个接收函数，打印接收的消息
def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard key: %d, data: %s", data.key, data.value)

def my_subscriber():
    rospy.init_node('my_subscriber', anonymous=True)
    #声明一个接收者
    rospy.Subscriber("my_topic", MyMessage, callback)
    #不断回调
    rospy.spin()

if __name__ == '__main__':
    try:
        my_subscriber()
    except rospy.ROSInterruptException:
        pass
